### PR TITLE
Initial message

### LINKED ISSUE #5160: Port `initialMessage` property for CatalogMember to v8

### PR DESCRIPTION
**Pull request by KeyboardSounds**
### What this PR does

Fixes #5160

Plus, the Notification refactor that nobody asked for!

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

File: lib/ModelMixins/AsyncMappableMixin.ts [hunk 1]
```diff
@@ -29,7 +30,8 @@ function AsyncMappableMixin<T extends Constructor<Model<MappableTraits>>>(
      * If the map items are already loaded or already loading, it will
      * return the existing promise.
      */
-    loadMapItems(): Promise<void> {
+    async loadMapItems(): Promise<void> {
+      await this.showInitialMessageIfRequired();
```
Comment (nf-s) [inline lib/ModelMixins/AsyncMappableMixin.ts]: I'm not sure this is the best place for this. What about putting something like:
```ts
if (ShowableMixin.isMixedInto(item) {
   await item.showInitialMessageIfRequired();
}
```
in https://github.com/TerriaJS/terriajs/blob/6c205d9fcb48857b098993e5723dedd5a5e24683/lib/Models/Workbench.ts#L195-L201

File: CHANGES.md [hunk 2]
```diff
@@ -12,6 +12,8 @@ Change Log
 * Add support for TableColumn quarterly dates in the format yyyy-Qx (eg 2020-Q1).
 * Fix region mapping feature highlighting.
 * Update clipboard to fix clipboard typing error.
+* Move notification state change logic from ViewState into new class NotificationState
+* Catalog items can now show a disclaimer or message before loading through specifying InitialMessageTraits
 * [The next improvement]
```
Comment (nf-s) [inline CHANGES.md]: Can you please add an entry about `ShowableMixin`

File: lib/ReactViewModels/ViewState.ts [hunk 3]
```diff
@@ -24,6 +22,7 @@ import {
   TourPoint
 } from "./defaultTourPoints";
 import DisclaimerHandler from "./DisclaimerHandler";
+import NotificationState from "./NotificationState";
```
Comment (nf-s) [inline lib/ReactViewModels/ViewState.ts]: There is still a `Notification` interface in this file

Comment (nf-s) [reply lib/ReactViewModels/ViewState.ts]: (at line 43 - it won't let me comment on it)

File: lib/ReactViewModels/ViewState.ts [hunk 4]
```diff
@@ -319,37 +318,15 @@ export default class ViewState {
     this.terria = terria;
 
     this._unsubscribeNotificationListener = terria.notification.addEventListener(
```
Comment (nf-s) [inline lib/ReactViewModels/ViewState.ts]: I think that if we're refactoring `Notification` stuff, we should probably get rid of the un-typed Cesium event handler in 

https://github.com/TerriaJS/terriajs/blob/20304b657d5e96766009f770a0c985f2345f8264/lib/Models/Terria.ts#L258

I know that `notificationState` should probably be in `ViewState`, but it needs to be visible to the entire codebase (especially the model layer). Could we maybe move it into `Terria.ts` and remove it from `ViewState`? As everywhere has access to the `terria` object, and then we can get rid of the `CesiumEvent` passing `Notifications` between `Terria` and `ViewState`.

File: lib/ReactViewModels/ViewState.ts (hunk 3 above)
Comment (nf-s) [reply lib/ReactViewModels/ViewState.ts]: https://github.com/TerriaJS/terriajs/blob/5b7d0155473d3537b7a2402a6fe40566080113cc/lib/ReactViewModels/ViewState.ts#L42-L53

File: lib/ReactViewModels/ViewState.ts [hunk 5]
```diff
@@ -639,14 +616,10 @@ export default class ViewState {
...
-  }
-
   hideMapUi() {
     return (
-      isDefined(this.getNextNotification()) &&
-      this.getNextNotification()!.hideUi
+      this.notificationState.currentNotification !== undefined &&
+      this.notificationState.currentNotification!.hideUi
     );
   }
```
Comment (nf-s) [inline lib/ReactViewModels/ViewState.ts]: Can we make this a `@computed` property?

File: lib/ReactViews/Notification/Notification.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  const viewState = props.viewState;
+  const notificationState = viewState?.notificationState;
+  const notification = notificationState?.currentNotification;
+
+  if (
+    viewState === undefined ||
+    notificationState === undefined ||
+    notification === undefined
+  ) {
+    return <></>;
```
Comment (nf-s) [inline lib/ReactViews/Notification/Notification.tsx:19]: I like returning `null` in these situations

Comment (nf-s) [reply lib/ReactViews/Notification/Notification.tsx:19]: probs doesn't matter though

File: lib/Traits/CatalogMemberTraits.ts [hunk 7]
```diff
@@ -4,6 +4,7 @@ import anyTrait from "./anyTrait";
 import LegendTraits from "./LegendTraits";
 import ModelTraits from "./ModelTraits";
 import objectArrayTrait from "./objectArrayTrait";
+import objectTrait from "./objectTrait";
```
Comment (nf-s) [inline lib/Traits/CatalogMemberTraits.ts]:
```suggestion
```

**Review by nf-s — CHANGES_REQUESTED** [review]
When I click on an item in the Data Catalog (using the default `simple` catalog)  - I get this

image

I also get this, when I click on `Marine water quality grids for the Great Barrier Reef region - monthly data` and then `number_of_valid_observations`

image

### EDIT

The change I suggest here - https://github.com/TerriaJS/terriajs/pull/5322#discussion_r592938193 - fixes the `@computed` error

Ahhh #5233 is going to be the end of me

File: lib/ReactViewModels/ViewState.ts (hunk 4 above)
Comment (nf-s) [reply lib/ReactViewModels/ViewState.ts]: This approach could also be used for the `terria.error` Cesium event handler

**Comment by KeyboardSounds** [comment]
TODO: Rewrite to observe changes to viewstate in an autorun, and run then, rather than in load

**Comment by nf-s** [comment]
Also, when you merge in `next` please take into account changes made to `Notification.jsx` here - https://github.com/TerriaJS/terriajs/pull/5413 

https://github.com/TerriaJS/terriajs/commit/08ea7a227a36902a00357710306225aa6f40e6cb#diff-a07b9cfd11367978cf4ae8e66b86910d059019e214c2a9bb5bf7403b2f95c617

**Review by nf-s — CHANGES_REQUESTED** [review]

File: lib/ModelMixins/AsyncMappableMixin.ts (hunk 1 above)
Comment (nf-s) [reply lib/ModelMixins/AsyncMappableMixin.ts]: We discussed this, you were right

File: lib/ReactViewModels/ViewState.ts (hunk 4 above)
Comment (nf-s) [reply lib/ReactViewModels/ViewState.ts]: Never mind, not needed right now

File: lib/Models/Terria.ts [hunk 8]
```diff
@@ -1632,10 +1635,14 @@ async function interpretHash(
 
         // Show warning messages if converted
         if (result.converted) {
-          terria.notification.raiseEvent({
+          // terria.notification.raiseEvent({
+          //   title: i18next.t("share.convertNotificationTitle"),
+          //   message: shareConvertNotification(result.messages)
+          // } as Notification);
```
Comment (nf-s) [inline lib/Models/Terria.ts]: stray comments

File: lib/ModelMixins/MappableMixin.ts [hunk 9]
```diff
@@ -7,8 +7,10 @@ import ImageryProvider from "terriajs-cesium/Source/Scene/ImageryProvider";
 import AsyncLoader from "../Core/AsyncLoader";
 import Constructor from "../Core/Constructor";
 import Model from "../Models/Model";
+import saveModelToJson from "../Models/saveModelToJson";
```
Comment (nf-s) [inline lib/ModelMixins/MappableMixin.ts]: this is unused

File: lib/ModelMixins/ShowableMixin.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import i18next from "i18next";
+import Property from "terriajs-cesium/Source/DataSources/Property";
+import { action, computed, observable } from "mobx";
+
+const t = i18next.t.bind(i18next);
+
+type ShowableModel = Model<ShowableTraits>;
+type InitialMessageJSON = {
+  [Property in keyof InitialMessageTraits]: any;
+};
```
Comment (nf-s) [inline lib/ModelMixins/ShowableMixin.ts]: This isn't used

File: lib/ModelMixins/ShowableMixin.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
+import Model from "../Models/Model";
+import Constructor from "../Core/Constructor";
+import ShowableTraits, { InitialMessageTraits } from "../Traits/ShowableTraits";
+import i18next from "i18next";
+import Property from "terriajs-cesium/Source/DataSources/Property";
```
Comment (nf-s) [inline lib/ModelMixins/ShowableMixin.ts]: This isn't used

**Review by nf-s — CHANGES_REQUESTED** [review]

**Review by nf-s — CHANGES_REQUESTED** [review]
Did we decide to get rid of `ShowableTraits`? I can't remember

**Review by nf-s — APPROVED** [review]

**Pull request merged**